### PR TITLE
Implemented TenureUpdatedEvent processing

### PR DESCRIPTION
**Pull request by mikemoody-amido**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/MTTL-1222

## Describe this PR

Process tenure updated event to update the tenure info on all person records that are listed in the tenure's household members collection.

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: PersonListener/Infrastructure/DateTimeExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+using System;
+
+namespace PersonListener.Infrastructure
+{
+    // TODO - Should probably go in a common NuGet package
+
+    public static class DateTimeExtensions
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply PersonListener/Infrastructure/DateTimeExtensions.cs:7]: Yes, hence the comment. I will add it to my ticket to move other stuff.

File: PersonListener/UseCase/TenureUpdatedUseCase.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                if (personTenure is null) throw new PersonMissingTenureException(thisPerson.Id, tenure.Id);
+
+                personTenure.AssetFullAddress = tenure.TenuredAsset.FullAddress;
+                personTenure.AssetId = tenure.TenuredAsset.Id.ToString();
+                personTenure.EndDate = tenure.EndOfTenureDate?.ToFormattedDateTime();
+                personTenure.PaymentReference = tenure.PaymentReference;
+                // personTenure.PropertyReference = tenure.PropertyReference; // TODO - property not yet available
+                personTenure.StartDate = tenure.StartOfTenureDate.ToFormattedDateTime();
+                personTenure.Type = tenure.TenureType.Description;
+                personTenure.Uprn = tenure.TenuredAsset.Uprn;
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply PersonListener/UseCase/TenureUpdatedUseCase.cs:59]: Should be. If they are null here then something is very wrong elsewhere so it's probably right that this should throw an exception.

File: PersonListener.Tests/UseCase/TenureUpdatedUseCaseTests.cs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            _mockGateway.Verify(x => x.GetPersonByIdAsync(It.IsAny<Guid>()), Times.Exactly(3));
+            _mockGateway.Verify(x => x.SavePersonAsync(It.IsAny<Person>()), Times.Exactly(3));
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task ProcessMessageAsyncTestAllPersonsUpdated(bool nullEndDate)
+        {
+            if (nullEndDate) _tenure.EndOfTenureDate = null;
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply PersonListener.Tests/UseCase/TenureUpdatedUseCaseTests.cs:187]: No. I just don't have specific comment lines that say that.

File: PersonListener/UseCase/TenureUpdatedUseCase.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+                updatedPersons.Add(thisPerson);
+            }
+
+            // #4 - Save all updated person records
+            foreach (var p in updatedPersons)
+                await _gateway.SavePersonAsync(p).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply PersonListener/UseCase/TenureUpdatedUseCase.cs:66]: Can do. It does make verifying the exception(s) raised in the tests much more difficult.

File: PersonListener/UseCase/TenureUpdatedUseCase.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+            // #2 - If there are no Household members we have nothing to do.
+            if (tenure.HouseholdMembers is null)
+                return;
+
+            // #3 - For each Household member update the tenure info on the person record
+            var updatedPersons = new List<Person>();
+            foreach (var hm in tenure.HouseholdMembers)
+            {
+                var thisPerson = await _gateway.GetPersonByIdAsync(hm.Id).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply PersonListener/UseCase/TenureUpdatedUseCase.cs:46]: Can do. It does make verifying the exception(s) raised in the tests much more difficult.

**Review by humulla — APPROVED** [review]

**Review by evak6979 — APPROVED** [review]

**Pull request merged**